File: traceharbor-battery-canary/src/main/java/io/traceharbor/batterycanary/monitor/feature/AbsMonitorFeature.java
package io.traceharbor.batterycanary.monitor.feature;

import android.content.pm.ApplicationInfo;
import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;

import io.traceharbor.batterycanary.monitor.BatteryMonitorCore;
import io.traceharbor.util.TraceHarborLog;

/**
 * @author Kaede
 * @since 2020/12/24
 */
public abstract class AbsMonitorFeature implements MonitorFeature {
    private static final String TAG = "TraceHarbor.battery.MonitorFeature";

    protected String getTag() {
        return TAG;
    }

    @SuppressWarnings("NotNullFieldNotInitialized")
    @NonNull
    protected BatteryMonitorCore mCore;

    @CallSuper
    @Override
    public void configure(BatteryMonitorCore monitor) {
        TraceHarborLog.i(getTag(), "#configure");
        this.mCore = monitor;
    }

    @CallSuper
    @Override
    public void onTurnOn() {
        TraceHarborLog.i(getTag(), "#onTurnOn");
    }

    @CallSuper
    @Override
    public void onTurnOff() {
        TraceHarborLog.i(getTag(), "#onTurnOff");
    }

    @CallSuper
    @Override
    public void onForeground(boolean isForeground) {
        TraceHarborLog.i(getTag(), "#onForeground, foreground = " + isForeground);
    }

    @CallSuper
    @WorkerThread
    @Override
    public void onBackgroundCheck(long duringMillis) {
        TraceHarborLog.i(getTag(), "#onBackgroundCheck, since background started millis = " + duringMillis);
    }

    protected boolean shouldTracing() {
        if (mCore.getConfig().isAggressiveMode) return true;
        return  0 != (mCore.getContext().getApplicationInfo().flags & ApplicationInfo.FLAG_DEBUGGABLE);
    }

    @Override
    public String toString() {
        return getTag();
    }
}

File: traceharbor-battery-canary/src/main/java/com/kernelflux/traceharbor/batterycanary/monitor/feature/AbsMonitorFeature.kt
package com.kernelflux.traceharbor.batterycanary.monitor.feature

import android.content.pm.ApplicationInfo
import androidx.annotation.CallSuper
import androidx.annotation.WorkerThread
import com.kernelflux.traceharbor.batterycanary.monitor.BatteryMonitorCore
import com.kernelflux.traceharbor.util.TraceHarborLog

/**
 * @author Kaede
 * @since 2020/12/24
 */
abstract class AbsMonitorFeature : MonitorFeature {
    protected open fun getTag(): String = TAG

    @JvmField
    protected var mCore: BatteryMonitorCore? = null

    protected val core: BatteryMonitorCore
        get() = checkNotNull(mCore) { "Monitor core is not configured yet" }

    @CallSuper
    override fun configure(monitor: BatteryMonitorCore) {
        TraceHarborLog.i(getTag(), "#configure")
        mCore = monitor
    }

    @CallSuper
    override fun onTurnOn() {
        TraceHarborLog.i(getTag(), "#onTurnOn")
    }

    @CallSuper
    override fun onTurnOff() {
        TraceHarborLog.i(getTag(), "#onTurnOff")
    }

    @CallSuper
    override fun onForeground(isForeground: Boolean) {
        TraceHarborLog.i(getTag(), "#onForeground, foreground = $isForeground")
    }

    @CallSuper
    @WorkerThread
    override fun onBackgroundCheck(duringMillis: Long) {
        TraceHarborLog.i(getTag(), "#onBackgroundCheck, since background started millis = $duringMillis")
    }

    protected fun shouldTracing(): Boolean {
        if (core.getConfig().isAggressiveMode) return true
        return 0 != (core.getContext().applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE)
    }

    override fun toString(): String = getTag()

    private companion object {
        private const val TAG = "TraceHarbor.battery.MonitorFeature"
    }
}

